taraxa/state/contracts/slashing/precompiled/slashing_contract.go

```diff
@@ -0,0 +1,428 @@
+
+	// Vrf sortition
+	VrfSortition VrfPbftSortition
+
+	// Signature
+	Signature [65]byte
+}
+
+func (self *Vote) GetRlp(include_sig bool) []byte {
+	rlp := rlp.MustEncodeToBytes(self)
+	if include_sig {
+		return rlp
+	}
+
+	return rlp[:len(rlp)-len(self.Signature)]
+}
+
+func (self *Vote) GetHash() *common.Hash {
+	return keccak256.Hash(self.GetRlp(false))
+}
+
+func NewVote(vote_rlp []byte) Vote {
+	var vote Vote
+	rlp.MustDecodeBytes(vote_rlp, &vote)
+
+	return vote
+}
+
+// Main contract class
+type Contract struct {
+	cfg Config
+
+	// current storage
+	storage contract_storage.StorageWrapper
+	// delayed storage for PBFT
+	read_storage Reader
+
+	// ABI of the contract
+	Abi abi.ABI
+	evm *vm.EVM
+
+	// Iterable map of malicious validators
+	malicious_validators contract_storage.AddressesIMap
+
+	// validator address -> list of proof of his malicious behaviour
+	validators_proofs map[common.Address]*ProofsIMap
+
+	// Double voting malicious behaviour proofs
+	double_voting_proofs DoubleVotingProofs
+}
+
+// Initialize contract class
+func (self *Contract) Init(cfg Config, storage contract_storage.Storage, read_storage Reader, evm *vm.EVM) *Contract {
+	self.cfg = cfg
+	self.storage.Init(slashing_contract_address, storage)
+	self.read_storage = read_storage
+	self.malicious_validators.Init(&self.storage, field_malicious_validators)
+	self.double_voting_proofs.Init(&self.storage, field_double_voting_proofs)
+	self.Abi, _ = abi.JSON(strings.NewReader(slashing_sol.TaraxaSlashingClientMetaData))
+	self.evm = evm
+	return self
+}
+
+// Register this precompiled contract
+func (self *Contract) Register(registry func(*common.Address, vm.PrecompiledContract)) {
+	defensive_copy := *slashing_contract_address
+	registry(&defensive_copy, self)
+}
+
+// Calculate required gas for call to this contract
+func (self *Contract) RequiredGas(ctx vm.CallFrame, evm *vm.EVM) uint64 {
+	method, err := self.Abi.MethodById(ctx.Input)
+	if err != nil {
+		return 0
+	}
+
+	switch method.Name {
+	case "commitDoubleVotingProof":
+		return CommitDoubleVotingProofGas
+	case "isJailed":
+		return IsJailedGas
+	case "getJailInfo":
+		return getJailInfoGas
+	case "getMaliciousValidators":
+		validators_count := uint64(self.malicious_validators.GetCount())
+		return validators_count * SlashingGetMethodGas
+	case "getDoubleVotingProofs":
+		// First 4 bytes is method signature !!!!
+		input := ctx.Input[4:]
+		var args slashing_sol.ValidatorArg
+		if err := method.Inputs.Unpack(&args, input); err != nil {
+			// args parsing will fail also during Run() so the tx wont get executed
+			return 0
+		}
+
+		proofs_count := uint64(self.getValidatorProofsList(&args.Validator).GetCount())
+		return proofs_count * SlashingGetMethodGas
+	default:
+	}
+
+	return DefaultSlashingMethodGas
+}
+
+// Should be called on each block commit
+func (self *Contract) CommitCall(read_storage Reader) {
+	// TODO: ClearCache breaks the tests for some reason ???
+	//defer self.storage.ClearCache()
```

Comment:
this breaks the tests, contract isJailed function does not work anymore... dont know why ???